p2p/base_reactor.go

```diff
@@ -160,64 +167,87 @@ func (br *BaseReactor) OnStop() {
 	br.cancel()
 }
 
-// DefaultProcessor unmarshalls the message and calls Receive on the reactor.
-// This preserves the sender's original order for all messages.
-func DefaultProcessor(impl Reactor) func(context.Context, <-chan UnprocessedEnvelope) error {
+// ProcessorWithReactor unmarshalls the message and calls Receive on the reactor.
+// This preserves the sender's original order for all messages and supports panic recovery with peer disconnection.
+func ProcessorWithReactor(impl Reactor, baseReactor *BaseReactor) func(context.Context, <-chan UnprocessedEnvelope) {
 	implChannels := impl.GetChannels()
 
 	chIDs := make(map[byte]proto.Message, len(implChannels))
 	for _, chDesc := range implChannels {
 		chIDs[chDesc.ID] = chDesc.MessageType
 	}
-	return func(ctx context.Context, incoming <-chan UnprocessedEnvelope) error {
+	return func(ctx context.Context, incoming <-chan UnprocessedEnvelope) {
 		for {
 			select {
 			case <-ctx.Done():
-				return nil
+				return
 			case ue, ok := <-incoming:
 				if !ok {
 					// this means the channel was closed.
-					return nil
+					return
 				}
-				mt := chIDs[ue.ChannelID]
 
-				if mt == nil {
-					return fmt.Errorf("no message type registered for channel %d", ue.ChannelID)
-				}
-
-				msg := proto.Clone(mt)
+				// Process message with panic recovery for individual peer
+				process := func(ue UnprocessedEnvelope) error {
+					defer func() {
+						if r := recover(); r != nil {
+							disconnectPeer(baseReactor, ue.Src, fmt.Sprintf("panic in reactor: %v", r), impl.String())
+						}
+					}()
```

Comment:
the main catch is here, previously (in all existing networks!!) comet processed things serially and has a different catch in the connection since it doesn't protect against nil messages in other spots 😑 

the rest below is minor refactors